hypothesis-python/tests/conjecture/test_ir_tree.py

```diff
@@ -0,0 +1,151 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Copyright the Hypothesis Authors.
+# Individual contributors are listed in AUTHORS.rst and the git log.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+from hypothesis.internal.conjecture.data import IRTree, IRTreeLeaf
+from hypothesis.internal.floats import SMALLEST_SUBNORMAL
+from hypothesis.internal.intervalsets import IntervalSet
+
+from tests.conjecture.common import fresh_data
+
+
+def test_copy_leaf():
+    leaf = IRTreeLeaf(
+        ir_type="integer",
+        value=10,
+        kwargs={"min_value": 0, "max_value": 100, "weights": None, "shrink_towards": 0},
+        was_forced=False,
+        depth=1,
+        index_in_parent=3,
+    )
+    assert leaf == leaf.copy()
+
+
+def test_copy_leaf_with_value():
+    leaf = IRTreeLeaf(
+        ir_type="integer",
+        value=10,
+        kwargs={"min_value": 0, "max_value": 100, "weights": None, "shrink_towards": 0},
+        was_forced=False,
+        depth=1,
+        index_in_parent=3,
+    )
+    leaf2 = leaf.copy(with_value=20)
+    assert leaf != leaf2
+    assert leaf2.value == 20
+
+
+def test_copy_empty_ir_tree():
+    ir_tree = IRTree()
+    assert ir_tree.root is None
+    assert ir_tree == ir_tree.copy()
+    assert ir_tree.copy().root is None
+
+
+def test_copy_ir_tree():
+    data = fresh_data()
+    data.draw_boolean(p=0.75)
+    data.start_example(42)
+    data.draw_integer()
+    data.stop_example()
+
+    assert data.ir_tree == data.ir_tree.copy()
```

Comment:
Stateful test time for this and the below!

(although keeping the last as "here's exactly what we expect" seems very valuable)